File: DataStructures/DisjointSet/python/DisjointSet.py
import sys

class DisjointSet:
	def __init__(self, n):
		self.n = n
		self.rank = [0] * self.n
		self.parent = [0] *self.n
		for i in range(n):
			self.parent[i] = i

	def find(self, x):
		if(self.parent[x] != x):
			self.parent[x] = self.find(self.parent[x])
		return self.parent[x]
	
	def _union(self, u, v):
		uset = self.find(u)
		vset = self.find(v)

		if(uset == vset):
			return

		if(self.rank[uset] < self.rank[vset]):
			self.parent[uset] = vset
		elif(self.rank[uset] > self.rank[vset]):
			self.parent[vset] = uset;
		else:
			self.parent[vset] = uset;
			self.rank[uset] += 1;

if __name__ == "__main__":

	sys.stdin = open('../input.txt', 'r')
	sys.stdout = open('../output.txt', 'w')

	n = int(sys.stdin.readline())

	ds = DisjointSet(n)

	for line in sys.stdin:

		if line[-1] == '\n':
			opt = line[:-1].split(" ")
		else:
			opt = line.split(" ")
		
		if opt[0] == "union":
			u, v = int(opt[1]), int(opt[2])
			ds._union(u, v)
		elif opt[0] == "find":
			u, v = int(opt[1]), int(opt[2])
			if(ds.find(u) == ds.find(v)):
				print(u,", ",v," belongs to same set.", end = "\n")
			else:
				print(u,", ",v," belongs to different sets.", end = "\n")

File: DataStructures/DisjointSet/c/DisjointSet.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <stdbool.h>

struct DisjSet{
	int *rank;
	int *parent;
	int n;
};

struct DisjSet* create(int n){
	struct DisjSet* ds = (struct DisjSet*)malloc(sizeof(struct DisjSet));
	ds -> n = n;
	ds -> rank = (int*)calloc(n, sizeof(int));
	ds -> parent = (int*)malloc(sizeof(int) * n);

	int i;
	for(i=0; i<n; i++){
		ds->parent[i] = i;
	}
	return ds;
}
int find(struct DisjSet* ds, int x){
	if(ds->parent[x] != x){
		ds->parent[x] = find(ds, ds->parent[x]);
	}
	return ds->parent[x];
}
void _union(struct DisjSet* ds, int u, int v){
	int uset = find(ds, u);
	int vset = find(ds, v);

	if(uset == vset)
		return;

	if(ds->rank[uset] < ds->rank[vset])
		ds->parent[uset] = vset;
	else if(ds->rank[uset] > ds->rank[vset])
		ds->parent[vset] = uset;
	else{
		ds->parent[vset] = uset;
		ds->rank[uset] += 1;
	}
}
int main(){

	FILE* fp = freopen("../input.txt", "r", stdin);
	freopen("../output.txt", "w", stdout);

	int n;
	scanf("%d",&n);

	struct DisjSet* ds = create(n);

	while(!feof(fp)){

		char opt[20]; 
		fgets(opt, 20, fp); // it reads string from a file, one line at a time

		if(!strncmp(opt,"union",strlen("union"))){ 
			char choice[10];
			int u, v;
			sscanf(opt, "%s %d %d" , choice, &u, &v);

			// u -> v
			_union(ds, u, v);
		}
		else if(!strncmp(opt,"find",strlen("find"))){ 
			char choice[10];
			int u, v;
			sscanf(opt, "%s %d %d" , choice, &u, &v);

			// u -> v
			if(find(ds, u) == find(ds, v)){
				printf("%d, %d belongs to same set\n",u,v);
			}
			else{
				printf("%d, %d belongs to different sets\n",u,v);
			}
		}
	}

	fclose(fp); 

}